CRM/Core/I18n/Form.php

```diff
@@ -61,13 +61,35 @@ public function buildQuickForm() {
     $dao->query($query, FALSE);
     $dao->fetch();
 
-    // we want TEXTAREAs for long fields and INPUTs for short ones
-    $this->_structure[$table][$field] == 'text' ? $type = 'textarea' : $type = 'text';
+    // get html type and attributes for this field
+    $widgets = CRM_Core_I18n_SchemaStructure::widgets();
+    $widget = $widgets[$table][$field];
+
+    // attributes
+    $attributes = array('class' => '');
+    if (isset($widget['rows'])) {
+      $attributes['rows'] = $widget['rows'];
+    }
+    if (isset($widget['cols'])) {
+      $attributes['cols'] = $widget['cols'];
+    }
+    $required = !empty($widget['required']);
 
     $languages = CRM_Core_I18n::languages(TRUE);
     foreach ($this->_locales as $locale) {
-      $this->addElement($type, "{$field}_{$locale}", $languages[$locale], array('class' => 'huge huge12' . ($locale == $tsLocale ? ' default-lang' : '')));
-      $this->_defaults["{$field}_{$locale}"] = $dao->$locale;
+      $name = "{$field}_{$locale}";
+      if ($locale == $tsLocale) {
+        $attributes['class'] .= ' default-lang';
+      }
+      if ($widget['type'] == 'RichTextEditor') {
+        $attributes['class'] .= ' collapsed';
+        $this->add('wysiwyg', $name, $languages[$locale], $attributes, $required);
```

Comment:
Maybe slightly cleaner would be for this line to say `$widget['type'] = 'wysiwyg'`. Then you don't need the next line to be wrapped in else.

Of course it would be even cleaner to have just 1 name for widget types... but we can't have everything.